### PR TITLE
Spins

### PR DESCRIPTION
**Pull request by SA-22C**
quick spins implementation, probably will require tweaking when it comes to calculating the degrees traversed.

### CODE REVIEW

File: sound/hybrid_font.h [hunk 1]
```diff
@@ -186,7 +191,13 @@ class HybridFont : public SaberBase {
...
     float rss = sqrtf(gyro_slope.z * gyro_slope.z + gyro_slope.y * gyro_slope.y) * (M_PI / 180);
     float swing_speed = fusor.swing_speed();
-    if (swing_speed > swingThreshold) {
+    uint32_t now = micros();
+    uint32_t delta_micros = now - last_swing_micros_;
+    last_swing_micros_ = now;
+    float delta = delta_micros * 0.000001;
+    angle_ += 0.98 * gyro.len() * delta;
+    angle_ += 0.02 * fusor.accel().len();
+      if (swing_speed > swingThreshold) {
```
Comment (profezzorn) [inline sound/hybrid_font.h:200]: Stupid github won't let me put the comment in the right place....
Anyways, I think we need to zero out angle_ when we're not swinging, which would be at the other branch of this if statement.

Comment (profezzorn) [inline sound/hybrid_font.h:200]: Why was this indented?

File: props/saber_sa22c_buttons.h [hunk 2]
```diff
@@ -279,7 +279,7 @@ class SaberSA22CButtons : public PropBase {
 #else
   // 2 and 3 button
   case EVENTID(BUTTON_AUX, EVENT_FIRST_SAVED_CLICK_SHORT, MODE_ON):
-  case EVENTID(BUTTON_AUX, EVENT_SECOND_CLICK_SHORT, MODE_ON):
+  case EVENTID(BUTTON_AUX, EVENT_SECOND_SAVED_CLICK_SHORT, MODE_ON):
```
Comment (profezzorn) [inline props/saber_sa22c_buttons.h:282]: This doesn't belong in this PR.

File: sound/hybrid_font.h [hunk 3]
```diff
@@ -186,7 +191,13 @@ class HybridFont : public SaberBase {
     // Radians per second per second
     float rss = sqrtf(gyro_slope.z * gyro_slope.z + gyro_slope.y * gyro_slope.y) * (M_PI / 180);
     float swing_speed = fusor.swing_speed();
-    if (swing_speed > swingThreshold) {
+    uint32_t now = micros();
+    uint32_t delta_micros = now - last_swing_micros_;
+    last_swing_micros_ = now;
+    float delta = delta_micros * 0.000001;
+    angle_ += 0.98 * gyro.len() * delta;
```
[the remark this answers is not in the document]
Comment (SA-22C) [reply sound/hybrid_font.h:198]: I did some testing and this stabilized the gyro length so that it didn't drift up while just sitting still.

File: sound/hybrid_font.h [hunk 4]
```diff
@@ -186,7 +191,13 @@ class HybridFont : public SaberBase {
...
+    angle_ += 0.02 * fusor.accel().len();
```
[the remark this answers is not in the document]
Comment (SA-22C) [reply sound/hybrid_font.h:199]: Same as the .098 stuff, combining the two seems to stabilize the gyro reading.

File: sound/hybrid_font.h [hunk 5]
```diff
@@ -207,11 +218,28 @@ class HybridFont : public SaberBase {
               swing_player_ = PlayPolyphonic(&SFX_swing);
             }
             swinging_ = true;
+          } else {
+            if (angle_ > font_config.ProffieOSSpinDegrees) {
+              if (SFX_spin) {
```
[the remark this answers is not in the document]
Comment (SA-22C) [reply sound/hybrid_font.h:223]: In my testing the swing sounds seemed to blend in too much with the hum pairs, so they were the least desirable accent spins in a fallback scenario.  Should we just fail back to swing if spin is not present?

[the remark this answers is not in the document]
Comment (SA-22C) [reply sound/hybrid_font.h:223]: I've changed it to just the swing_speed from fusor, will test.

File: sound/hybrid_font.h [hunk 6]
```diff
@@ -186,7 +191,13 @@ class HybridFont : public SaberBase {
     // Radians per second per second
     float rss = sqrtf(gyro_slope.z * gyro_slope.z + gyro_slope.y * gyro_slope.y) * (M_PI / 180);
     float swing_speed = fusor.swing_speed();
-    if (swing_speed > swingThreshold) {
+    uint32_t now = micros();
+    uint32_t delta_micros = now - last_swing_micros_;
```
[the remark this answers is not in the document]
Comment (SA-22C) [reply sound/hybrid_font.h:195]: resolved.

File: sound/hybrid_font.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (SA-22C) [reply sound/hybrid_font.h]: ok, removed both and just using fusor.swing_speed(), works fine.

File: sound/hybrid_font.h [hunk 7]
```diff
@@ -207,11 +218,28 @@ class HybridFont : public SaberBase {
...
+              if (SFX_spin) {
+                swing_player_ = PlayPolyphonic(&SFX_spin);
+              } else if (SFX_slsh) {
+                swing_player_ = PlayPolyphonic(&SFX_slsh);
+              } else if (SFX_swng) {
+                swing_player_ = PlayPolyphonic(&SFX_swng);
+              } else {
+                swing_player_ = PlayPolyphonic(&SFX_swing);
+              }
+              angle_ = 0;
```
[the remark this answers is not in the document]
Comment (SA-22C) [reply sound/hybrid_font.h]: done

File: sound/hybrid_font.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (SA-22C) [reply sound/hybrid_font.h]: removed this and just using swing_speed()

[the remark this answers is not in the document]
Comment (SA-22C) [reply sound/hybrid_font.h]: done

File: sound/hybrid_font.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (SA-22C) [reply sound/hybrid_font.h:195]: done

File: sound/hybrid_font.h (hunk 1 above)
Comment (SA-22C) [reply sound/hybrid_font.h]: done

Comment (SA-22C) [reply sound/hybrid_font.h]: error indenting, corrected.

File: props/saber_sa22c_buttons.h (hunk 2 above)
Comment (SA-22C) [reply props/saber_sa22c_buttons.h]: removed

File: sound/hybrid_font.h [hunk 8]
```diff
@@ -207,17 +218,35 @@ class HybridFont : public SaberBase {
...
+          PlayMonophonic(&SFX_spin, &SFX_hum);
+          angle_ -= font_config.ProffieOSSpinDegrees;
+        }
       }
       float swing_strength = std::min<float>(1.0, swing_speed / swingThreshold);
       SetSwingVolume(swing_strength, 1.0);
     } else if (swing_speed <= font_config.ProffieOSSwingLowerThreshold) {
       swinging_ = false;
       swing_player_.Free();
+      angle_ = 0;
```
Comment (profezzorn) [inline sound/hybrid_font.h:249]: By doing this, any drift we have should go away when we're not swinging.

File: sound/hybrid_font.h [hunk 9]
```diff
@@ -186,6 +191,12 @@ class HybridFont : public SaberBase {
     // Radians per second per second
     float rss = sqrtf(gyro_slope.z * gyro_slope.z + gyro_slope.y * gyro_slope.y) * (M_PI / 180);
     float swing_speed = fusor.swing_speed();
+    uint32_t now = micros();
+    uint32_t delta_micros = now - last_swing_micros_;
+    last_swing_micros_ = now;
+    if (delta_micros > 1000000) delta_micros = 1;
+    float delta = delta_micros * 0.000001;
+    angle_ += fusor.swing_speed() * delta;
```
Comment (profezzorn) [inline sound/hybrid_font.h:199]: Actually, we already have fuser.swing_speed() in a local variable "swing_speed".

File: sound/hybrid_font.h [hunk 10]
```diff
@@ -186,6 +191,12 @@ class HybridFont : public SaberBase {
     // Radians per second per second
     float rss = sqrtf(gyro_slope.z * gyro_slope.z + gyro_slope.y * gyro_slope.y) * (M_PI / 180);
     float swing_speed = fusor.swing_speed();
+    uint32_t now = micros();
+    uint32_t delta_micros = now - last_swing_micros_;
+    last_swing_micros_ = now;
+    if (delta_micros > 1000000) delta_micros = 1;
+    float delta = delta_micros * 0.000001;
```
Comment (profezzorn) [inline sound/hybrid_font.h:198]: It occurs to me that line 198 & 199 can be moved into the if statement below.
We don't need to do those calculations when we're not swinging.

File: sound/hybrid_font.h (hunk 8 above)
Comment (SA-22C) [reply sound/hybrid_font.h:245]: agreed, tested fine.

**Comment by SA-22C** [comment]
Ok, removed fallbacks and did requested cleanups/tweaks.

File: sound/hybrid_font.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (SA-22C) [reply sound/hybrid_font.h:223]: Ok, removed all fallbacks.
